#!/usr/bin/env python3
"""
Test client for the KillChat and HealthCheck endpoints
"""

import asyncio
import grpc
from grpc import aio
import chat_service_pb2
import chat_service_pb2_grpc


async def test_health_check():
    """Test the health check endpoint"""
    print("ğŸ©º Testing Health Check endpoint...")
    
    try:
        async with aio.insecure_channel('localhost:7000') as channel:
            stub = chat_service_pb2_grpc.AiOrchestratorStub(channel)
            
            request = chat_service_pb2.HealthCheckRequest()
            response = await stub.HealthCheck(request)
            
            print(f"âœ… Health Check Response:")
            print(f"   Status: {response.status}")
            print(f"   Message: {response.message}")
            print(f"   Timestamp: {response.timestamp}")
            print(f"   Details: {dict(response.details)}")
            
            return True
            
    except Exception as e:
        print(f"âŒ Health check failed: {e}")
        return False


async def test_kill_chat():
    """Test the kill chat endpoint"""
    print("\nğŸ”ª Testing Kill Chat endpoint...")
    
    try:
        async with aio.insecure_channel('localhost:7000') as channel:
            stub = chat_service_pb2_grpc.AiOrchestratorStub(channel)
            
            # Test killing a non-existent session
            request = chat_service_pb2.KillChatRequest(
                session_id="test_session_123",
                reason="Testing kill functionality"
            )
            response = await stub.KillChat(request)
            
            print(f"ğŸ“‹ Kill Chat Response (non-existent session):")
            print(f"   Success: {response.success}")
            print(f"   Message: {response.message}")
            
            # Test killing with no session ID
            request = chat_service_pb2.KillChatRequest(
                reason="Testing default session kill"
            )
            response = await stub.KillChat(request)
            
            print(f"\nğŸ“‹ Kill Chat Response (default session):")
            print(f"   Success: {response.success}")
            print(f"   Message: {response.message}")
            
            return True
            
    except Exception as e:
        print(f"âŒ Kill chat test failed: {e}")
        return False


async def main():
    """Main test function"""
    print("ğŸš€ Starting gRPC endpoint tests...\n")
    
    # Test health check
    health_ok = await test_health_check()
    
    # Test kill chat
    kill_ok = await test_kill_chat()
    
    print(f"\nğŸ“Š Test Results:")
    print(f"   Health Check: {'âœ… PASS' if health_ok else 'âŒ FAIL'}")
    print(f"   Kill Chat: {'âœ… PASS' if kill_ok else 'âŒ FAIL'}")
    
    if health_ok and kill_ok:
        print("\nğŸ‰ All tests passed!")
    else:
        print("\nâš ï¸ Some tests failed. Make sure the server is running.")


if __name__ == "__main__":
    asyncio.run(main())